### PR TITLE
Add script for checking titles

### PR DESCRIPTION
**Pull request by janetkuo**
Follow up #2007. Update the script to check titles also, and rename the script since it now checks more than TOC. @devin-donnelly 

##### Docs missing titles:
- [x] docs/api-reference/extensions/v1beta1/definitions.md
- [x] docs/api-reference/extensions/v1beta1/operations.md
- [x] docs/api-reference/v1/definitions.md
- [x] docs/api-reference/v1/operations.md
- [x] docs/getting-started-guides/docker.md
- [x] docs/getting-started-guides/stackpoint.md
- [x] docs/tutorials/stateless-application/hello-minikube.md
- [x] docs/user-guide/configmap/README.md
- [x] docs/user-guide/downward-api/README.md
- [ ] docs/user-guide/kubectl/kubectl_certificate.md
- [ ] docs/user-guide/kubectl/kubectl_certificate_approve.md
- [ ] docs/user-guide/kubectl/kubectl_options.md
- [ ] docs/user-guide/kubectl/kubectl_certificate_deny.md
- [ ] docs/user-guide/kubectl/kubectl_cluster-info_dump.md
- [ ] docs/user-guide/kubectl/kubectl_completion.md
- [ ] docs/user-guide/kubectl/kubectl_config_delete-cluster.md
- [ ] docs/user-guide/kubectl/kubectl_config_delete-context.md
- [ ] docs/user-guide/kubectl/kubectl_config_get-clusters.md
- [ ] docs/user-guide/kubectl/kubectl_config_get-contexts.md
- [ ] docs/user-guide/kubectl/kubectl_rollout_status.md
- [ ] docs/user-guide/kubectl/kubectl_cp.md
- [ ] docs/user-guide/kubectl/kubectl_set.md
- [ ] docs/user-guide/kubectl/kubectl_set_image.md
- [ ] docs/user-guide/kubectl/kubectl_create_deployment.md
- [ ] docs/user-guide/kubectl/kubectl_set_resources.md
- [ ] docs/user-guide/kubectl/kubectl_create_quota.md
- [ ] docs/user-guide/kubectl/kubectl_taint.md
- [ ] docs/user-guide/kubectl/kubectl_create_service_nodeport.md
- [ ] docs/user-guide/kubectl/kubectl_top.md
- [ ] docs/user-guide/kubectl/kubectl_create_secret_tls.md
- [ ] docs/user-guide/kubectl/kubectl_top_node.md
- [ ] docs/user-guide/kubectl/kubectl_create_service.md
- [ ] docs/user-guide/kubectl/kubectl_create_service_clusterip.md
- [ ] docs/user-guide/kubectl/kubectl_create_service_loadbalancer.md
- [ ] docs/user-guide/kubectl/kubectl_top_pod.md
- [ ] docs/user-guide/pods/_viewing-a-pod.md
- [ ] docs/user-guide/walkthrough/index.md
- [ ] docs/user-guide/walkthrough/k8s201.md
- [ ] docs/user-guide/simple-yaml.md

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/2151)

### CODE REVIEW

**Comment by janetkuo** [comment]
```
$ ./verify-docs-format.sh
Error: docs/api-reference/extensions/v1beta1/definitions.md doesn't have a proper title defined!
Error: docs/api-reference/extensions/v1beta1/operations.md doesn't have a proper title defined!
Error: docs/api-reference/v1/definitions.md doesn't have a proper title defined!
Error: docs/api-reference/v1/operations.md doesn't have a proper title defined!
Error: docs/getting-started-guides/docker.md doesn't have a proper title defined!
Error: docs/getting-started-guides/stackpoint.md doesn't have a proper title defined!
Error: docs/tutorials/stateless-application/hello-minikube.md doesn't have a proper title defined!
Error: docs/user-guide/configmap/README.md doesn't have a proper title defined!
Error: docs/user-guide/downward-api/README.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_certificate.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_certificate_approve.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_options.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_certificate_deny.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_cluster-info_dump.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_completion.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_config_delete-cluster.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_config_delete-context.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_config_get-clusters.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_config_get-contexts.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_rollout_status.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_cp.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_set.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_set_image.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_create_deployment.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_set_resources.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_create_quota.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_taint.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_create_service_nodeport.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_top.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_create_secret_tls.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_top_node.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_create_service.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_create_service_clusterip.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_create_service_loadbalancer.md doesn't have a proper title defined!
Error: docs/user-guide/kubectl/kubectl_top_pod.md doesn't have a proper title defined!
Error: docs/user-guide/pods/_viewing-a-pod.md doesn't have a proper title defined!
Error: docs/user-guide/walkthrough/index.md doesn't have a proper title defined!
Error: docs/user-guide/walkthrough/k8s201.md doesn't have a proper title defined!
Error: docs/user-guide/simple-yaml.md doesn't have a proper title defined!
Error: docs/sitemap.md doesn't have a proper title defined!
Found 40 files without titles.
```

File: verify-docs-format.sh [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  fi
+
+  # 2. Title check:
+  #    Check they have a proper title.
+  #    Skip checking files in skip_title_check.txt
+  #    Title should start with "title:" and can have several spaces between
+  #    non-space content. For example, "title:", " title:", or "title: " isn't valid, 
+  #    but "title: abc" and "title:    def ghi" are both valid.
+  if ! grep -q "${file}" skip_title_check.txt; then
+    if ! grep -q "^title: *[^ ]" ${file}; then
```
Comment (pwittrock) [inline verify-docs-format.sh]: This accepts tabs as non-space.  Should this be "^title:\s*[^\s]" instead?

**Comment by pwittrock** [comment]
Looks good.  Do we need to fix the docs before merging this?

File: verify-docs-format.sh (hunk 1 above)
Comment (janetkuo) [inline verify-docs-format.sh]: Fixed

**Comment by janetkuo** [comment]
> Do we need to fix the docs before merging this?

Yeah, will wait for docs to be fixed before merging

File: verify-docs-format.sh [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  fi
+
+  # 2. Title check:
+  #    Check they have a proper title.
+  #    Skip checking files in skip_title_check.txt
+  #    Title should start with "title:" and can have several spaces/tabs between
+  #    non-space/tab content. 
+  #    For example, "title:", " title: abc", and "title:" aren't valid, 
+  #    but "title: abc", "title:def" and "title:    def ghi" are both valid.
+  if ! grep -q "${file}" skip_title_check.txt; then
```
Comment (pwittrock) [inline verify-docs-format.sh]: WDYT of adding everything to skip title check to stop the bleeding, and remove them incrementally as we fix them?

File: skip_title_check.txt [hunk 3]
```diff
@@ -0,0 +1,2 @@
+# Put files you want to skip "titles:" check here:
```
Comment (devin-donnelly) [inline skip_title_check.txt]: "title:" check; we don't want to confuse people as to what the attribute is actually called.

File: verify-docs-format.sh [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    # abc/index.md should point to abc, not abc/index
+    path=${path%%index}
+    if ! grep -q "${path}" _data/*.yml; then
+      echo "Error: ${file} doesn't have an entry in the table of contents under _data/*.yml" 
+      no_entry=true
+      no_entry_counter=$[no_entry_counter+1]
+    fi
+  fi
+
+  # 2. Title check:
```
Comment (devin-donnelly) [inline verify-docs-format.sh]: Specify that the "title:" attribute needs to be inside the markdown header?

(E.g. inside the --- --- at the top of the doc?)

**Comment by devin-donnelly** [comment]
@janetkuo , great! Can you handle the kubectl docs, like last time?

**Comment by janetkuo** [comment]
> Can you handle the kubectl docs, like last time?

Sure!

**Comment by devin-donnelly** [comment]
@janetkuo Non-kubectl docs updated. Most of these are either empty autogenerated stuff or deprecated/empty docs, so I added them to the skip-check file. A couple needed title updates and I added those.

**Comment by devin-donnelly** [comment]
@janetkuo Also, we'll need to update the script to account for the new-style reference documentation that @pwittrock has added.

**Comment by janetkuo** [comment]
@devin-donnelly I update the script to be able to skip folders. I didn't update titles for those auto-generated docs/user-guide/kubectl/kubectl_xxx.md since they already work (since their first heading is what we'd put in the title, we don't need to add their titles explicitly)

We can now merge this.

**Pull request merged**